text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,774 @@
+
+Here one function uses `become` the other `return`, this is another potential source of confusion. This mutual recursion
+would eventually overflow the stack. As mutual recursion can also happen across more functions, `become` needs to be
+used consistently in all functions if TCE is desired. (Maybe it is also possible to create a lint for these
+use cases as well.)
+
+```rust
+fn foo(n: i32) {
+    // oops, we forgot become ..
+    return bar(n); // or alternatively: `bar(n)`
+}
+
+fn bar(n: i32) {
+    become foo(n);
+}
+```
+
+### Pitfall 4: Caller location aka [`panic::Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller) and `#[track_caller]`
+
+Since `#[track_caller]` adds an argument, `#[track_caller]` functions can only tail call other `#[track_caller]` functions. So the following example will not compile.
+
+```rust
+use std::panic::Location;
+
+/// Returns the [`Location`] at which it is called.
+#[track_caller]
+fn get_caller_location() -> &'static Location<'static> {
+    Location::caller()
+}
+
+/// Returns a [`Location`] from within this function's definition.
+fn get_just_one_location() -> &'static Location<'static> {
+    become get_caller_location();
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+Implementation of this feature requires checks that all prerequisites to guarantee TCE are fulfilled.
+These checks are:
+
+- The `become` keyword is only used in place of `return`. The intent is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.
+- The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intent is to ensure a matching ABI. Note that lifetimes may differ as long as they pass borrow checking, see [below](#return-type-coercion) for specifics on the return type.
+- The stack frame of the calling function is reused, this also implies that the function is never returned to. The required checks to ensure this is possible are: no borrows of local variables are passed to the called function (passing local variables by copy/move is ok since that doesn't require the local variable to continue existing after the call), and no further cleanup is necessary. These checks can be done by using the borrow checker as already described in the [section](#difference) showing the difference between `return` and `become` above.
+
```

Comment:
I think we should add something like

> - Additionally there are some restrictions on the caller context/callee signature, see below

since we have more restrictions below, like banning `c_variadic` or explicitrly banning closures, etc.

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/0b61f079d8eb0a8caf216d1937a770d5fafb188e